### ISSUE TITLE
Is it possible to make color/custom more custom?

### ISSUE THREAD

**Issue by bralexander** [nglviewer/ngl#796]
Hi, I want to color the protein with machine learning data from a csv. I have been messing with the codepen unsuccessfully thus-far. My question is, in the color assigning function, would it be possible to do something like this:

this.atomColor = function(atom, mlData) {
  if (atom.residueIndex == mlData.residueIndex && mlData.prob > 0.5) {
    return 0xFF000
  }
}
*there definitely needs to be a for statement there too, but it gives you the gist

Where the mlData would come from a local csv file that is parsed into an object.
How would you go about passing the csv-object(mlData) into the color assignment function?
Ultimately, I'd like to make a whole new colormaker scheme and use the web app, but I've just been working on the basics for now.

**Comment by fredludlow** [comment]
Yes, sort of like that - you can't change the way that atomColor is called, but if you've already got the mlData the you make a closure. 

At the moment you need to add a new scheme per molecule to ensure the data gets incorporated, (as you're doing above)

Based on this example: https://github.com/nglviewer/ngl/blob/master/examples/scripts/color/custom.js and untested :)
```javascript
var mlData = getMlData()

var schemeId = NGL.ColormakerRegistry.addScheme(function (params) {
  this.atomColor = function (atom) {
    var prob = mlData[atom.residueIndex]
    if (prob == undefined) {
      return 0xFFFFFF // white 
    } else if (prob <= 0.5) {
      return 0x00FF00 // green
    } else {
      return 0xFF0000 // red
    }
  }
})

stage.loadFile('data://3dqb.pdb').then(function (o) {
  o.addRepresentation('cartoon', { color: schemeId })
  o.autoView()
})
```

If you want to use the inbuilt scales, you can extend NGL.Colormaker (see for example the implementation of BfactorColormaker: https://github.com/nglviewer/ngl/blob/master/src/color/bfactor-colormaker.ts).

If getMlData is asynchronous you need to do something like `Promise.all([getMlData(), stage.loadFile()]).then(...`

If you want to do this in a more general way (so users can just select "my-ML-scheme" from the list) then it's currently a bit hacky. I've done this (don't have the code to hand so can't remember quite how) but I think I extended Colormaker: this means your colormaker instance has a reference to the structure and I have a helper function which provides the structure -> data mapping. It's not very satisfactory though and if the data comes asynchronously you need to handle that appropriately.

**Comment by bralexander** [comment]
Thanks Fred! Got it working. My code is pretty hacky, because the csv I'm using starts on residue 9, so I'll have to make sure the csv's are indexed properly. I'm also going to add some more user inputs so it is more customizable. 

I'll also probably want to have a default csv that uploads with the pdb, so you can actually see something when it loads. Would this file need to be stored in the NGL data:// directory?

**Comment by fredludlow** [comment]
You can get the residue number for an atom directly if that helps? (Perhaps not if your CSV file doesn't include residue numbers :)
```javascript
this.atomColor = function(atom) {
  atom.resno // Residue number
  atom.residueIndex // 0-based index
}
```

`atom` is an instance of `AtomProxy`: https://nglviewer.org/ngl/api/classes/atomproxy.html.

> I'll also probably want to have a default csv that uploads with the pdb, so you can actually see something when it loads. Would this file need to be stored in the NGL data:// directory?

You can put it there, or anywhere else you can Ajax/Fetch from. The webapp example just sets up the data:// prefix for convenience: https://github.com/nglviewer/ngl/blob/master/examples/webapp.html#L32
`NGL.DatasourceRegistry.add("data", new NGL.StaticDatasource("../data/"))`

**Comment by bralexander** [comment]
I got the codepen working well now, and I now want to try working with it as an example in the web app. I have the web app running locally, and I've made a file that extends Colormaker. I have that file in src/color, example/scripts/color, and I updated the build/scriptlist.json (which I'm pretty sure is where the example tab in the web app gets its list?). I also see two files in build/js/src/color. I haven't been able to see my file in the example list on the web app. Am I missing something? Any suggestions?
